build.gradle.kts

```diff
@@ -188,78 +187,113 @@ nexusPublishing {
+						// this prevents the corresponding warning (and keeps working on Java 8)
+						// IF YOU ADD MORE OPTIONS; CONSIDER REPLACING `-XX:+IgnoreUnrecognizedVMOptions WITH A CONDITIONAL
+						jvmArgs(
+								"-XX:+IgnoreUnrecognizedVMOptions",
+								"--add-opens=java.base/java.util=ALL-UNNAMED")
+
+					}
+				}
 			}
-			compileClasspath += sourceSets.main.get().output
-			runtimeClasspath += sourceSets.main.get().output
-		}
-	}
-	project(":demo") {
-		sonarqube {
-			isSkipProject = true
 		}
-	}
-	// Adds all dependencies of main to demo sourceSet
-	configurations["demoImplementation"].extendsFrom(configurations.implementation.get())
-	// Ensures JUnit 5 engine is available to demo at runtime
-	configurations["demoRuntimeOnly"].extendsFrom(configurations.testImplementation.get())
 
-	compileJava {
-		options.encoding = "UTF-8"
-	}
-
-	compileTestJava {
-		options.encoding = "UTF-8"
-	}
+		val demoTests by registering(JvmTestSuite::class) {
+			dependencies { implementation(project) }
 
-	test {
-		configure<JacocoTaskExtension> {
-			isEnabled = !experimentalBuild
-		}
-		testLogging {
-			setExceptionFormat("full")
-		}
-		useJUnitPlatform()
-		filter {
-			includeTestsMatching("*Tests")
+			sources { java { srcDir("src/demo/java") } }
+			targets {
+				all {
+					testTask.configure {
+						shouldRunAfter(test)
+						filter {
+							includeTestsMatching("*Demo")
+						}
+					}
+				}
+			}
 		}
-		systemProperty("java.util.logging.manager", "org.apache.logging.log4j.jul.LogManager")
-		// `EnvironmentVariableExtension` uses reflection to change environment variables;
-		// this prevents the corresponding warning (and keeps working on Java 8)
-		// IF YOU ADD MORE OPTIONS; CONSIDER REPLACING `-XX:+IgnoreUnrecognizedVMOptions WITH A CONDITIONAL
-		jvmArgs(
-				"-XX:+IgnoreUnrecognizedVMOptions",
-				"--add-opens=java.base/java.util=ALL-UNNAMED")
-	}
 
-	testing {
-		suites {
-			val test by getting(JvmTestSuite::class) {
+		supportedJUnitVersions.split(",").filter { it != junitVersion }.forEach {
+			val testSuite = register("testWithJUnit$it", JvmTestSuite::class) {
 				useJUnitJupiter()
-			}
-			val demoTests by registering(JvmTestSuite::class) {
-				dependencies { implementation(project) }
-
-				sources { java { srcDir("src/demo/java") } }
-				targets { all { testTask.configure {
-					shouldRunAfter(test)
-					filter {
-						includeTestsMatching("*Demo")
+
+				dependencies {
+					implementation(project)
+					pioneerImplementation.forEach { dependency ->
+						implementation(dependency)
+					}
+					pioneerTestImplementation.forEach {dependency ->
+						implementation(dependency)
+					}
+					implementation(project.dependencies.platform("org.junit:junit-bom:$it"))
+				}
+
+				sources {
+					java {
+						srcDir("src/test/java")
+					}
+					resources {
+						srcDir("src/test/resources")
+					}
+				}
+
+				targets {
+					all {
+						testTask.configure {
```

Comment:
It looks like you could also specify the Java version to run the test here:
```kotlin
javaLauncher.set(javaToolchains.launcherFor {
    languageVersion.set(JavaLanguageVersion.of(javaVersion))
})
```

Then it might not be necessary anymore to run the complete build with different JDK versions (unless that is desired).

Reply:
looks good, i added that